docs/pxlOptions.md
# pxlNav Docs - pxlOptions
### Options & User Settings objects
This is the rundown of the default options object you pass to the `pxlNav` constructor

### `pxlUserSettings` object -
```
import { pxlNav, pxlUserSettings, pxlOptions } from './pxlNav.###.js';

// -- -- --

// To be safe, clone the imported `pxlUserSettings` object
const userSettings = Object.assign({}, pxlUserSettings);

// Standing height in units; any camera in your room's FBX will override this height once loaded
userSettings['height']['standing'] = 1.75; 

// Max step height in units
userSettings['height']['stepSize'] = 5; 

// Overall movement rate scalar
userSettings['movement']['scalar'] = 1.0; 

// Max movement speed
userSettings['movement']['max'] = 10.0; 

// Easing rate between Step() calls
userSettings['movement']['easing'] = 0.55; 

// Bounce magnitude in units
userSettings['headBounce']['height'] = 0.3; 

// Bounce rate per Step()
userSettings['headBounce']['rate'] = 0.025; 

// When move key is pressed, the ease into bounce; `bounce * ( boundInf + easeIn )`
userSettings['headBounce']['easeIn'] = 0.03; 

// When move key is let go, the ease back to no bounce; `bounce * easeOut`
userSettings['headBounce']['easeOut'] = 0.95; 

// Jump impulse force applied to the player while holding the jump button
userSettings['jump']['impulse'] = 0.75; 

// Max influence of holding the jump button on current jump; in seconds
userSettings['jump']['holdMax'] = 2.85; 

// Delay between jumps when holding the jump button
userSettings['jump']['repeatDelay'] = 0.08;

// Units per Step() per Step()
userSettings['gravity']['UPS'] = 0.3; 

// Max gravity rate
userSettings['gravity']['Max'] = 15.5;

// - - -

// To be safe, clone the imported `pxlOptions` object
let pxlNavOptions = Object.assign( {}, pxlOptions );
pxlNavOptions.userSettings = userSettings;

const pxlNavEnv = new pxlNav( pxlNavOptions, "Your Project Title", "StartingRoomName", ["StartingRoomName","AnotherRoomName"] );

```



### Defaults -

```
export const pxlUserSettings = {
  'height' : {
      'standing' : 1.75,
      'stepSize' : 5
    },
  'movement' : {
      'scalar' : 1.0,
      'max' : 10.0,
      'easing' : 0.55
    },
  'headBounce' : {
      'height' : 0.3,
      'rate' : 0.025,
      'easeIn' : 0.03,
      'easeOut' : 0.95
    },
  'jump' : {
      'impulse' : 0.75,
      'holdMax' : 2.85,
      'repeatDelay' : 0.08
    },
  'gravity' : {
      'UPS' : 0.3, // Units per Step()
      'Max' : 15.5
    },
  };
```


```
export const pxlOptions = {
  'verbose' : pxlEnums.VERBOSE_LEVEL.NONE,
  'fps' : {
    'PC' : 60,
    'Mobile' : 30
  },
  'subTickCalculations' : false,
  'pxlRoomRoot' : "./pxlRooms",
  'pxlAssetRoot' : "./pxlAssets",
  'showOnboarding' : true,
  'staticCamera' : false,
  'autoCamera' : false,
  'userSettings' : Object.assign({}, pxlUserSettings),
  'antiAliasing' : pxlEnums.ANTI_ALIASING.LOW,
  'collisionScale' : {
    'gridSize' : 50,
    'gridReference' : 1000
  },
  'shadowMapBiasing' : pxlEnums.SHADOW_MAP.BASIC,
  'LoadEnvAssetFile' : false,
  'skyHaze' : pxlEnums.SKY_HAZE.OFF,
  'loaderPhrases' : ['...loading the pixels...']
}
```

ChangeLog.md
# pxlNav Change Log :: 0.0.19 - 0.0.20
---------------------

 - Changed the example name from `FieldEnvironment` to `OutletEnvironment`
<br/>&nbsp;&nbsp; - It wasn't really a field anymore hah

---

 - `Enums.js` + `Options.js` now has `VERBOSE_LEVEL.DEBUG` for dumping all logging data
<br/>&nbsp;&nbsp; - You can still use the verbose flag when loading FBXs from your pxlRoom without being in debug mode
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - `this.pxlFile.loadRoomFBX( this, null, null, true );`

 - `pxlNav.js` + `Environment.js` + `QualityController.js` + `Devices.js` better asset and post-process handling
<br/>&nbsp;&nbsp; - It no longer loads all the assets for the optional post-process passes
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - Still needs updates to handle loading during a run, but can be set through `pxlOptions.postProcessPasses`

---

 - `Utils.js` bug preventing texture mods from being applied has been fixed

---

 - `FileIO.js` has the begining's of spot light implementation
<br/>&nbsp;&nbsp; - Not in the scope of this version, to be added soon.

 - `FileIO.js` now supports LODs in your instance objects in your scene FBX
<br/>&nbsp;&nbsp; - Support for N-number of LOD levels, simply add User-Details / Parameters -
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - `instLOD1` ; String of your `Instance` mesh name to use at this level
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - `instLOD1Dist` ; Distance to trigger this level
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - `instLOD1Skip` ; Verts count to skip over spawning the instance to the points of this mesh

<br/>&nbsp;&nbsp; - Simply update the `*` to the desired number LOD you are making the parameter for
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - `instLOD*` ; `instLOD*Dist` ; `instLOD*Skip` ; 

<br/>&nbsp;&nbsp; - With 2 methods of assigning your base instance mesh inst-to-mesh
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - Use a child mesh as the instance-base mesh to use as point locations
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - No child mesh and set `Skip` number to reduce how many verts as skipped when creating the LOD instance-base mesh. 

<br/>&nbsp;&nbsp; - For mesh LODs, put the LOD as children under your instance-to-mesh objects
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - Your object-mesh with the `Instance` parameter is the parent transform.
<br/>&nbsp;&nbsp; - Add a float User Detail of - `instLOD1Dist`, `instLOD2Dist`,  `instLOD[...]Dist`
<br/>&nbsp;&nbsp;&nbsp;&nbsp; - Float value is the distance the LOD level displays at


---

src/pxlNav/core/Options.js


import { pxlEnums } from './Enums.js';

export const pxlUserSettings = {
  'height' : {
      'standing' : 1.75, // Standing height in units; any camera in your room's FBX will override this height once loaded
      'stepSize' : 5 // Max step height in units
    },
  'movement' : {
      'scalar' : 1.0, // Overall movement rate scalar
      'max' : 10.0, // Max movement speed
      'easing' : 0.55 // Easing rate between Step() calls
    },
  'headBounce' : {
      'height' : 0.3, // Bounce magnitude in units
      'rate' : 0.025, // Bounce rate per Step()
      'easeIn' : 0.03, // When move key is pressed, the ease into bounce; `bounce * ( boundInf + easeIn )`
      'easeOut' : 0.95 // When move key is let go, the ease back to no bounce; `bounce * easeOut`
    },
  'jump' : {
      'impulse' : 0.75, // Jump impulse force applied to the player while holding the jump button
      'holdMax' : 2.85, // Max influence of holding the jump button on current jump; in seconds
      'repeatDelay' : 0.08 // Delay between jumps when holding the jump button
    },
  'gravity' : {
      'UPS' : 0.3, // Units per Step() per Step()
      'Max' : 15.5 // Max gravity rate
    },
  };


export const pxlOptions = {
  'verbose' : pxlEnums.VERBOSE_LEVEL.NONE,
  'fps' : {
    'PC' : 60,
    'Mobile' : 30
  },
  'subTickCalculations' : false,
  'pxlRoomRoot' : "./pxlRooms",
  'pxlAssetRoot' : "./pxlAssets",
  'showOnboarding' : true,
  'staticCamera' : false,
  'autoCamera' : false,
  'userSettings' : Object.assign({}, pxlUserSettings),
  'antiAliasing' : pxlEnums.ANTI_ALIASING.LOW,
  'collisionScale' : {
    'gridSize' : 50,
    'gridReference' : 1000
  },
  'shadowMapBiasing' : pxlEnums.SHADOW_MAP.BASIC,
  'LoadEnvAssetFile' : false,
  'skyHaze' : pxlEnums.SKY_HAZE.OFF,
  'loaderPhrases' : ['...loading the pixels...'],
  'postProcessPasses' : { // Enabling loads needed ./pxlAssets
    'roomGlowPass' : false,
    'mapComposerWarpPass' : false,
    'chromaticAberrationPass' : false,
    'lizardKingPass' : false,
    'starFieldPass' :  false,
    'crystallinePass' : false
  }
}